File: 2115_Find_All_Possible_Recipes_from_Given_Supplies.cpp
class Solution {
public:
    vector<string> findAllRecipes(vector<string>& recipes, vector<vector<string>>& ingredients, vector<string>& supplies) {
        int n = recipes.size();
        vector<int> indeg(n, 0);
        vector<string> ret;
        unordered_map<string, vector<int>> mp; // item, recipe index (ri)

        for (int i = 0; i < n; i++)
        {
            for (auto ingredient : ingredients[i]) 
            {
                indeg[i]++;
                mp[ingredient].push_back(i); 
            }
        }

        queue<string> q;
        // all supplies have indeg 0 by default
        for (auto &supply : supplies) 
            q.push(supply);

        while (!q.empty())
        {
            auto t = q.front(); 
            q.pop();

            for (int ri : mp[t])
            {
                indeg[ri]--;
                if (indeg[ri] == 0)
                {
                    ret.push_back(recipes[ri]);
                    q.push(recipes[ri]);
                }
            }
        }

        return ret;
    }
};

File: py/2115_Find_All_Possible_Recipes_from_Given_Supplies.py
from collections import deque, defaultdict 
from typing import List

class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        n = len(recipes)
        indeg = [0] * n
        ret = [] 

        mp = defaultdict(list) 

        for i in range(n):
            for ingredient in ingredients[i]:
                indeg[i] += 1
                mp[ingredient].append(i)

        q = deque(supplies) 

        while q: 
            t = q.popleft() 

            for ri in mp[t]: 
                indeg[ri] -= 1
                if indeg[ri] == 0:
                    ret.append(recipes[ri])
                    q.append(recipes[ri])
        return ret
